### PR TITLE
Add anti aliasing rendering hints regardless of system default

### PR DESCRIPTION
**Pull request by vinceh121**
Anti aliasing rendering hints can be applied even when system defaults are null.

Additionally, MaterialMenuUI and MaterialTextFieldUI now use anti aliased renders.

### CODE REVIEW

File: src/main/java/mdlaf/components/menu/MaterialMenuUI.java
Comment (vincenzopalazzo) [inline src/main/java/mdlaf/components/menu/MaterialMenuUI.java:71]: This is not necessary because MaterialMenuUI it is a fake component and under the hood is a MateriaMenuItem, and the library support this function already.

The superclass of MaterialMenuUI contains the following code

```
    public void paint(Graphics g, JComponent c) {
        paintMenuItem(g, c, checkIcon, arrowIcon,
                      selectionBackground, selectionForeground,
                      defaultTextIconGap);
    }
```

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Hello @vinceh121, I really like this change, good catch, thanks for contributing to the library. I added a couple of comments.

P.S: I need to change my JDK with the JDK 8 and make more tests because the antialiasing is an evident problem on JDK, I will test also on it.

File: src/main/java/mdlaf/components/menu/MaterialMenuUI.java
Comment (vinceh121) [reply src/main/java/mdlaf/components/menu/MaterialMenuUI.java:71]: This confuses me because `BasicMenuItemUI#paintMenuItem` will not call `MaterialMenuItemUI#paint` and so not use its anti-aliased Graphics.
This is evidenced with a simple test run:
no AA in MaterialMenuUI
with AA in MaterialMenuUI

File: src/main/java/mdlaf/components/textfield/MaterialTextFieldUI.java
Comment (vinceh121) [reply src/main/java/mdlaf/components/textfield/MaterialTextFieldUI.java]: Change made in a038c63 with no troublesome variable name

File: src/main/java/mdlaf/components/menu/MaterialMenuUI.java
Comment (vincenzopalazzo) [reply src/main/java/mdlaf/components/menu/MaterialMenuUI.java:71]: Mh! This is a very good catch. Thanks for pointing me to this problem. This night I will make more tests in JDK 8 and after I will accept this PR.

**Comment by vincenzopalazzo** [comment]
I'm back @vinceh121,

This is the result of the JDK 8 on Linux



Unfortunately, this is a JDK limit and requires a very deep change inside the library to support this "feature".

Do you have any idea to fix the anti-aliasing on JDK 8?

**Comment by vinceh121** [comment]
It seems that applying the Graphics hint `KEY_FRACTIONALMETRICS` with value `VALUE_FRACTIONALMETRICS_OFF` fixes the problem.

**Comment by zanderson9** [comment]
> It seems that applying the Graphics hint KEY_FRACTIONALMETRICS with value VALUE_FRACTIONALMETRICS_OFF fixes the problem.

This is great news! at the JMARS team, we are waiting to have this "cursor-type-ahead" issue fixed in JDK 8.

**Comment by vincenzopalazzo** [comment]
@vinceh121 I'm sorry if I'm late to make a answer, but I miss my self inside a refactoring of the `MaterialDrawingUtils`, because it has very bad behavior, and I'm sorry for that.

Each time that someone puts somethings inside a HashMap it needs to check if there is already something inside the hash map.

What it is mean that? that the call to `awt.font.desktophints` get the rendering configuration, it means if we put directly some value inside the has map we can override some default configuration, and it is very bad.

I taked a little bit of time to refactor the class and [here there is the result](https://github.com/vincenzopalazzo/material-ui-swing/blob/refactoring_rednder_utils/src/main/java/mdlaf/utils/MaterialDrawingUtils.java).

>It seems that applying the Graphics hint KEY_FRACTIONALMETRICS with value VALUE_FRACTIONALMETRICS_OFF fixes the problem.

I added your suggestion and it worked, thanks.

@vinceh121 If you agree I will rebase your PR with the new version of MaterialDrawingUtils and I can make another test with this solution.

**Comment by vinceh121** [comment]
> If you agree I will rebase your PR with the new version of MaterialDrawingUtils and I can make another test with this solution.

I agree, everything looks good to me :+1:

**Comment by vincenzopalazzo** [comment]
Thanks, @vinceh121 this night I will proceed with the rebase and make some test with the different app and in the next few days I will join the PR if there are no problems.

**Comment by vincenzopalazzo** [comment]
I'm back @vinceh121,

I and @zanderson9 made a test with [JMars]() on Macbook Air (laptop), Retina display, JDK 8 and we noted this side effect

[antialias-cusor.pdf](https://github.com/vincenzopalazzo/material-ui-swing/files/5879522/antialias-cusor.pdf)

Before all, I won't see that the JDK on some computer version doesn't help us to make a cool effect but this "side effect" shows why to work the anti-aliasing is dangerous, because required a lot of testing and it depends from the JDK.


The source code at the moment is on the following branch https://github.com/vincenzopalazzo/material-ui-swing/tree/rebase_pt

**Comment by vinceh121** [comment]
Hi @vincenzopalazzo,
To be clear the problem here is the same "cursor-type-ahead"? If so this confuses me because the component used here seems to me to be either a `JTextArea` or a `JEditorPane`, however in the `rebase_pt` branch those don't have anti aliasing applied to them.

Also the refactored `MaterialDrawingUtils` doesn't set hints that already have a default value set, but wouldn't this cause contradicting hints? The problem might be linked to having part custom hints + part default macOS hints set.
https://github.com/vincenzopalazzo/material-ui-swing/blob/b00646b2d1aeff0d010125b738c386e296394a6f/src/main/java/mdlaf/utils/MaterialDrawingUtils.java#L85

**Comment by vincenzopalazzo** [comment]
Hi @vinceh121 

> To be clear the problem here is the same "cursor-type-ahead"?

yes.

>If so this confuses me because the component used here seems to me to be either a JTextArea or a JEditorPane, however in the rebase_pt branch those don't have anti-aliasing applied to them.

The anti-aliasing was applied to JLabel and it could involve several components, but I'm not sure about that and I'm not enough time to walk inside the JDK hierarchy at the moment, but somethings changed.

>Also the refactored MaterialDrawingUtils doesn't set hints that already have a default value set, but wouldn't this cause contradicting hints? The problem might be linked to having part custom hints + part default macOS hints set.

For sure the problem might be the conflict of propriety inside the table, or might the null hash table returned by the following code, and this mean that the table gets the default table conf, but it contains the wrong result for the JDK 8. So in cases, the method returns a null value the table need to be null

https://github.com/vincenzopalazzo/material-ui-swing/blob/b00646b2d1aeff0d010125b738c386e296394a6f/src/main/java/mdlaf/utils/MaterialDrawingUtils.java#L62

P.S: The JDK 8 has a different library under the hood from the modern JDK to managed the graphics and fonts in the Java application, for this reason, I'm a little worried to touch the null/empty table returned from the previous method because it can save one platform but can destroy somethings else.

But at the moment I'm not able to prove that and the error could be inside my refactoring, for this reason, I will take more time to review this PR before merging it.

If you have an additional test, please share it at any time.

Thanks.

File: src/main/java/mdlaf/utils/MaterialDrawingUtils.java
Comment (vincenzopalazzo) [inline src/main/java/mdlaf/utils/MaterialDrawingUtils.java:66]: This code needs to be refactored, you can support something like that

```
    Graphics2D g2d = (Graphics2D) g;
        g2d.addRenderingHints(hints);
        return g2d;
```

The comment after return is a very bad thing in my opinion.

File: src/main/java/mdlaf/utils/MaterialDrawingUtils.java
Comment (vincenzopalazzo) [inline src/main/java/mdlaf/utils/MaterialDrawingUtils.java:53]: Can you apply your suggestion to switch this part to OFF?

File: src/main/java/mdlaf/utils/MaterialDrawingUtils.java
Comment (vincenzopalazzo) [inline src/main/java/mdlaf/utils/MaterialDrawingUtils.java:51]: I'm not completely happy with this solution, because could have happened a collision between propriety.

But this is a problem that I want to solve and I'm happy to work around with you to choose the best solution to use in this case.

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Hello @vinceh121, I work around a little bit on this PR and I can confirm that this PR works well with the full HD monitor, but @zanderson9 has a problem with the Retina display.

I guess that this is not a problem of this PR but is things that live in the JDK 8, but  I need to work around to support the JDK 8 because the library is used with JDK 8 in big projects.

In conclusion, I want proposed to split the PR into two-part, one that adds the anti-aliasing rendering to JMenu and JTextField.

And only in a second moment, we think about how to set up the hints inside the map in the correct way, because with the method in this PR I think is too aggressive, but the method that I propose in some branches is also wrong because sometimes the propriety to put inside the map can have some collision and print a wrong result

File: src/main/java/mdlaf/utils/MaterialDrawingUtils.java
Comment (vinceh121) [reply src/main/java/mdlaf/utils/MaterialDrawingUtils.java:53]: Shouldn't it be `KEY_FRACTIONALMETRICS` that needs to be switched off?

Comment (vincenzopalazzo) [reply src/main/java/mdlaf/utils/MaterialDrawingUtils.java:53]: Oh you have right, I point out the wrong line. I'm so sorry :-D So, yes  it is `KEY_FRACTIONALMETRICS`

**Review by vincenzopalazzo — APPROVED** [review]

**Comment by vinceh121** [comment]
Hi @vincenzopalazzo,

About the problem with the Retina display, I would assume it would come from the JRE's default desktop hints on macOS turning on font fractional metrics and it isn't being overridden to off because of the logic of your refactor.

> I want proposed to split the PR into two-part, one that adds the anti-aliasing rendering to JMenu and JTextField.

Then this would already be done, right?

> And only in a second moment, we think about how to set up the hints inside the map in the correct way, because with the method in this PR I think is too aggressive, but the method that I propose in some branches is also wrong because sometimes the propriety to put inside the map can have some collision and print a wrong result

When it comes to setting render hints, it seems to me that the safest bet would be to statically override a copy of the default desktop hints map so we can be sure that the required hints for the LAF are there (turning on AA, font AA, and off fractional metrics, etc...) and use said copy for Material renders; this likely wouldn't cause trouble to other libraries since the default map wouldn't be changed. In which case collisions shouldn't be a problem as long as we force the problematic hints.

**Comment by vincenzopalazzo** [comment]
Hello @vinceh121,

>About the problem with the Retina display, I would assume it would come from the JRE's default desktop hints on macOS turning on font fractional metrics and it isn't being overridden to off because of the logic of your refactor.

Yes, this makes sense, but I want to point out some comment on my code inside the PR

```
    private static Map<RenderingHints.Key, Object> putMissedHints(Map<RenderingHints.Key, Object> hints) {
        if (hints == null) {
            return defaultHints;
        }
        for (Map.Entry<RenderingHints.Key, Object> entry : defaultHints.entrySet()) {
            if (!hints.containsKey(entry.getKey())) {
                hints.put(entry.getKey(), entry.getValue());
            }
        }
        return hints;
    }
```

This code makes your suggestion, but in a not aggressive way, and with it I assuming that the JDK knows the details when it returns the hints about the correct way. I'm saying that because I'm assuming that the look and feel inside the JDK don't make this override operation, but it use the default configuration from the JDK. So, for this reason, if the JDK said for the key `KEY_FRACTIONALMETRICS` the correct value is `VALUE_FRACTIONALMETRICS_ON` I'm not the body to override this key.

The point that my code is not able to prove is when the hints map is different to null the logic inside the if put some pair (Key1, value1) that is a bad choice with the (Key2, value2).
But I repeat my self, I'm assuming that if the JDK returns the value `VALUE_FRACTIONALMETRICS_ON` for the key `KEY_FRACTIONALMETRICS`  it is correct for the host machine.

At the moment I'm have not available the a correct configuration to prove that your work well on the retina display and I don't wan revert the merge if there is some problem. So, for me looks good if it resolve the bug on the retina display and thank to work on it :-D.

I'm ping @zanderson9  that has the correct configuration of the host machine to reproduce the bug, and it has the last word on this PR.

**Comment by vincenzopalazzo** [comment]
Hello @vinceh121,

I'm here again with a couple of tests that you can see inside this [Drive directory](https://drive.google.com/drive/folders/1hqpJb_xbsv1dqxUcdXkvU0cHOSYjoHCy?usp=sharing)

I make respectively 2 different test, subdivided in the following directory

- jar1: In this test, I ran this PR without my change and I received the following result

cursor2 (1)

- jar2: in this test, your PR with my change, but I removed them for loop inside to check if a variable already exists inside the map, so, jar two contains a refactoring of your PR, and the result is the following

textfield-test2 (1)

The environment is a Macbook Air with Java 8 and Retina Display.

So at the moment, I can not accept this PR as is. But, there are a couple of good points, such as the anti-aliased on MaterialMenuUI, but not on the JTextFieldComponent.

So, if you want you can open another PR with this change and I can merge right now, for this PR, I will work with @zanderson9 to catch the bug.

Thanks for the patient.

p.s: A big thanks to @zanderson9 that help me to make the tests.

**Comment by vincenzopalazzo** [comment]
Closing now this PR, thanks for contributing.

**Pull request closed without merging**
